py_wtf/repository.py

```diff
@@ -34,28 +30,36 @@ def _load_from_disk(self, key: ProjectName) -> None:
         index_file = self._index_file(key)
         index_contents = index_file.read_bytes()
         proj = converter.loads(index_contents, Project)
-        self._cache[key] = proj
+        self._cache[key].set_result(proj)
 
     def _save(self, project: Project) -> None:
         name = ProjectName(project.name)
-        if name in self._cache:
+        if self._cache[name].done():
             return
 
-        self._cache[name] = project
+        self._cache[name].set_result(project)
         index_file = self._index_file(name)
         index_file.write_text(converter.dumps(project))
 
-    def get(
+    async def get(
         self,
         key: ProjectName,
-        factory: Callable[[ProjectName], Iterable[Project]],
+        factory: Callable[[ProjectName], AsyncIterable[Project]],
     ) -> Project:
+        if key in self._cache:
+            return await self._cache[key]
+
         try:
-            return self[key]
-        except KeyError:
+            self._load_from_disk(key)
+            return await self._cache[key]
+        except OSError:
             pass  # continued below
-        projects = factory(key)
-        for project in projects:
+
+        # It's important to force the creation of this future before we await
+        # to make sure everyone awaits on the same future and so avoid duplicating work
+        fut = self._cache[key]
+
+        async for project in factory(key):
             self._save(project)
 
-        return self[key]
+        return await fut
```

Comment:
This is still a bit dodgy; if `factory(key)` never `async yield`s the `project` with `.name == key` then this `await` will block indefinitely (or until another poor coroutine happens to fulfil the future).

I'll fix this and add a test for it in a followup PR

Reply:
(y) At the very least on the happy path this will actually work like well-behaved futures!